dali/python/nvidia/dali/shared_batch.py

```diff
@@ -0,0 +1,259 @@
+        # used to identify shared memory chunks in the communication between processes
+        self.mem_chunk_id = mem_chunk_id
+        self.shm_chunk = shared_mem.SharedMem.allocate(capacity)
+        self.capacity = capacity
+
+    def resize(self, new_capacity):
+        self.shm_chunk.resize(new_capacity, trunc=True)
+        self.capacity = new_capacity
+
+    def close(self):
+        # Shared memory handle and mapping will be freed automatically when shm pybind wrapper gets
+        # garbage collected anyway, but you can be so nice to free them as soon as you know
+        # you won't use them
+        self.shm_chunk.close()
+
+
+def _to_numpy(sample):
+    if isinstance(sample, np.ndarray):
+        return sample
+    if types._is_mxnet_array(sample):
+        return sample.asnumpy()
+    if types._is_torch_tensor(sample):
+        if sample.device.type != 'cpu':
+            raise TypeError("GPU tensors are not supported")
+        return sample.numpy()
+    elif isinstance(sample, tensors.TensorCPU):
+        return np.array(sample)
+    raise TypeError(
+        "Unsupported callback return type. Expected numpy array, pytorch or mxnet cpu tensors, "
+        "or list or tuple of them.")
+
+
+def _apply_to_sample(func, sample, *args, nest_with_sample=0):
+    """Apply to a sample traversing the nesting of the data (tuple/list).
+
+    Parameters
+    ----------
+    func : callable
+        Function to be applied to every sample data object
+    sample : sample object or any nesting of those in tuple/list
+        Representation of sample
+    nest_with_sample: int
+        Specify how many consecutive (additional) arguments have the same level of nesting
+        as the sample.
+    """
+    if isinstance(sample, (tuple, list,)):
+        # Check that all the samples have common nesting
+        for i in range(nest_with_sample):
+            assert len(args[i]) == len(sample)
+        nest_group = sample, *args[0:nest_with_sample]
+        scalar_args = args[nest_with_sample:]
+        return type(sample)(_apply_to_sample(func, *part, *scalar_args) for part in zip(*nest_group))
+    else:
+        # we unpacked all nesting levels, now is actual data:
+        return func(sample, *args)
+
+
+class SharedBatchWriter:
+    """SharedBatchWriter can serialize and write batch into given shared
+    memory chunk (``mem_batch``).
+    """
+
+    SAMPLE_ALIGNMENT = 128
+    BUFFER_ALIGNMENT = 4096
+
+    def __init__(self, mem_batch: SharedMemChunk, batch):
+        import_numpy()
+        self.mem_batch = mem_batch
+        self.data_size = 0
+        self.meta_data_size = 0
+        self._write_batch(batch)
+
+    def _prepare_samples_meta(self, indexed_samples):
+        """Calculate metadata and total size of data to be serialized"""
+        data_size = 0
+
+        def make_meta(np_array):
+            nonlocal data_size
+            offset = _div_ceil(data_size, self.SAMPLE_ALIGNMENT) * self.SAMPLE_ALIGNMENT
```

Comment:
nitpick:
```suggestion
            offset = _align_up(data_size, self.SAMPLE_ALIGNMENT)
```
where
```Python
def _align_up(x, alignment):
    return -(-x & (alignment-1))
```

Reply:
where `_align_up(1, 128) == -127`. 

I will use _div_ceil, I can wrap it in _align_up if you think it's better.